content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,95 @@
+So even if custom code has not changed, that does not mean after an API update it will keep working the same. 
+Lastly, because of the ubiquitous use of these steps throughout Pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoiding large global variable declaration files
+
+Having large variable declaration files can require large amounts of memory for little to no benefit, because the file is loaded for every Pipeline whether the variables are needed or not. Creating small variable files that contain only variables relevant to the current execution is recommended.
+It would be better to only load variables which are relevant to the current execution.
+
+=== Avoiding very large shared libraries
+
+Using large shared libraries in Pipelines requires checking out a very large file before the Pipeline can start and loading the same shared library per job that is currently executing, which can lead to increased memory overhead and slower execution time.
+
+== Answering additional FAQs
+
+=== Dealing with Concurrency in Pipelines
+
+Try not to share workspaces across multiple Pipeline executions or multiple distinct Pipelines. 
+This practice can lead to either unexpected file modification within each Pipeline or workspace renaming.
+
+Ideally, shared volumes/disks are mounted in a separate place and the files are copied from that place to the current workspace.
+Then when the build is done the files can be copied back if there were updates done.
+
+Build in distinct containers which create needed resources from scratch(cloud-type agents work great for this). 
+Building these containers will ensure that the build process begins at the start every time and is easily repeatable.
+If building containers will not work, disable concurrency on the Pipeline or use the Lockable Resources Plugin to lock the workspace when it is running so that no other builds can use it while it is locked.
+**WARNING**: Disabling concurrency or locking the workspace when it is running can cause Pipelines to become blocked when waiting on resources if those resources are arbitrarily locked.
+
+**Also, be aware that both of these methods have slower time to results of builds than using unique resources for each job**
+
+=== Using @NonCPS
+
+Pipeline code is CPS-transformed so that Pipelines are able to resume after a Jenkins restart. Some Groovy expressions do not work correctly as a result of this transformation. See http://jenkins.io/redirect/pipeline-cps-method-mismatches for more details and some examples of things that may be problematic. If necessary, you can use the @NonCPS annotation to disable the CPS transformation for a specific method whose body would not execute correctly if it were CPS-transformed. Just be aware that this also means the Groovy function will have to restart completely since it is not transformed.
```

Comment:
I would prefer to make the first reference to an acronym (like 'CPS') include the expansion of that acronym.  However, your opening text in the paragraph was so good at describing the value for the user that I delayed the explanation of the acronym into a later paragraph.  I copied the original text from [Kohsuke's 2014 blog post](https://jenkins.io/blog/2014/12/03/workflow-plugin-is-1-0/) and modified it to use current terminology.

You are welcome to reject the entire suggestion if you feel it is too dramatic a change.  I think the material is great, with or without this suggestion.

```suggestion
Pipeline code is CPS-transformed so that Pipelines are able to resume after a Jenkins restart.
That is, while the pipeline is running your script, you can shut down Jenkins or lose connectivity to an agent. 
When it comes back, Jenkins remembers what it was doing and your pipeline script resumes execution as if it were never interrupted.
A technique known as "link:https://en.wikipedia.org/wiki/Continuation-passing_style[continuation-passing style (CPS)]"  execution plays a key role in resuming Pipelines.

However, some Groovy expressions do not work correctly as a result of CPS transformation.
See link:http://jenkins.io/redirect/pipeline-cps-method-mismatches[Pipeline CPS method mismatches] for more details and some examples of things that may be problematic.

If necessary, you can use the `@NonCPS` annotation to disable the CPS transformation for a specific method whose body would not execute correctly if it were CPS-transformed.
Just be aware that this also means the Groovy function will have to restart completely since it is not transformed.
```